### PR TITLE
new speech recognition API

### PR DESCRIPTION
**Pull request by RaduDumitriu-Dev**
- adds speech recognition
- uses new unified message format
- adds speech recognition configuration class, w/ several paramaters
- at time of writing languageCode and encoding do not seem to produce any results and have been commented out. need clarification on these two during feedback

### CODE REVIEW

File: Scripts/Audio.cs [hunk 1]
```diff
@@ -183,4 +183,138 @@ public string output
...
+    {
+        public enum SpeechRecognitionService
+        {
+            Unified,
+            Google,
+            Aws,
+            Deepgram
+        }
+
+        public enum SpeechRecognitionLanguageCode
```
Comment (John-A-J) [inline Scripts/Audio.cs:198]: I think we should not be prescriptive with the language code and just accept any string directly. Different services allow different language codes and may also add or change accepted codes in the future

File: Scripts/Audio.cs [hunk 2]
```diff
@@ -183,4 +183,138 @@ public string output
...
+                    case SpeechRecognitionService.Unified:
+                        return "unified";
+                    case SpeechRecognitionService.Google:
+                        return "unified:google";
+                    case SpeechRecognitionService.Aws:
+                        return "unified:aws";
+                    case SpeechRecognitionService.Deepgram:
+                        return "unified:deepgram";
+                    default:
+                        Logger.LogError("Unknown audio format");
```
Comment (John-A-J) [inline Scripts/Audio.cs:230]:
```suggestion
                        Logger.LogError("Unrecognised SpeechRecognitionService");
```

File: Scripts/Events.cs [hunk 3]
```diff
@@ -26,6 +26,21 @@ public class ProblemEvent
         public string error;
     }
 
+    [Preserve]
+    [Serializable]
+    public class SpeechRecognitionResult
+    {
+        public int confidence;
+
+        public int durationInSeconds;
```
Comment (bensalilijames) [inline Scripts/Events.cs:35]: `int` -> `float`

File: Scripts/Audio.cs [hunk 4]
```diff
@@ -183,4 +183,138 @@ public string output
...
+        public int SampleRate => _sampleRate;
+
+        [SerializeField]
+        private SpeechRecognitionService _service = SpeechRecognitionService.Unified;
+
+        [SerializeField]
+        private SpeechOptions.Encoding _encoding = SpeechOptions.Encoding.Mp3;
+
+        [SerializeField]
+        private SpeechRecognitionLanguageCode _languageCode = SpeechRecognitionLanguageCode.EN_GB;
```
Comment (bensalilijames) [inline Scripts/Audio.cs:302]: default should be `en-US` to match server

File: Scripts/Audio.cs [hunk 5]
```diff
@@ -183,4 +183,138 @@ public string output
...
+                    case SpeechRecognitionService.Deepgram:
+                        return "unified:deepgram";
+                    default:
+                        Logger.LogError("Unknown audio format");
+                        return null;
+                }
+            }
+        }
+
+        public string Encoding
```
Comment (bensalilijames) [inline Scripts/Audio.cs:236]: I don't think we need `Encoding` as we control the microphone implementation so we know what it's going to be

File: Scripts/Audio.cs [hunk 6]
```diff
@@ -183,4 +183,138 @@ public string output
     }
 
 
+
+    [Serializable]
+    public class SpeechRecognitionOptions
```
Comment (bensalilijames) [inline Scripts/Audio.cs:188]: maybe put all the speech recognition stuff into a separate file to avoid cluttering `Audio`

**Review by bensalilijames — CHANGES_REQUESTED** [review]

**Review by bensalilijames — APPROVED** [review]
awesome, looks good! 👍

**Pull request merged**